src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+                if !character.is_alphanumeric() {
+                    is_data_alphanum = false;
+                }
+            }
+            if is_data_alphanum {
+                label.push_str(&"_");
+                label.push_str(&self.nodes[id.index].data.as_str());
+            }
+        }
+        dot::LabelText::LabelStr(label.into())
+    }
+}
+
+impl<'a, T, U> dot::GraphWalk<'a, NodeId, EdgeId> for Arena<T, U> {
+    fn nodes(&self) -> dot::Nodes<'a, NodeId> {
+        let mut nodeids: Vec<NodeId> = vec![];
+        for index in 0..self.nodes.len() {
+            nodeids.push(NodeId { index: index });
+        }
+        Cow::Owned(nodeids)
+    }
+
+    fn edges(&'a self) -> dot::Edges<'a, EdgeId> {
+        Cow::Owned(self.get_edges())
+    }
+
+    fn source(&self, e: &EdgeId) -> NodeId {
+        let &(s, _) = e;
+        NodeId { index: s }
+    }
+
+    fn target(&self, e: &EdgeId) -> NodeId {
+        let &(_, t) = e;
+        NodeId { index: t }
+    }
+}
+
+// Should have same output as lrpar::parser::Node<TokId>::pretty_print().
+impl<T, U> fmt::Display for Arena<T, U>
+    where T: fmt::Display,
+          U: fmt::Display
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.nodes.is_empty() {
+            return write!(f, "");
+        }
+        let mut formatted = String::new();
+        let mut stack: Vec<NodeId> = vec![NodeId { index: 0 }];
+        while !stack.is_empty() {
+            let id = stack.pop().unwrap();
+            let node = &self.nodes[id.index];
+            if node.is_leaf() {
+                // Terminal node.
+                formatted.push_str(&format!("{}\n", node));
+            } else {
+                // Non-terminal node.
+                formatted.push_str(&format!("{}\n", node));
+                let first_child = node.first_child.unwrap();
+                let mut current_child = node.last_child.unwrap();
+                loop {
+                    stack.push(current_child);
+                    if current_child == first_child {
+                        break;
+                    }
+                    match self.nodes[current_child.index].previous_sibling {
+                        None => break,
+                        Some(id) => current_child = id,
+                    };
+                }
+            }
+        }
+        write!(f, "{}", formatted)
+    }
+}
+
+// #[derive(Clone, Debug)]
+/// An AST node within an `Arena`.
+pub struct Node<T, U> {
+    parent: Option<NodeId>,
+    previous_sibling: Option<NodeId>,
+    next_sibling: Option<NodeId>,
+    first_child: Option<NodeId>,
+    last_child: Option<NodeId>,
+    /// The actual data which will be stored within the tree.
+    pub data: T,
+    /// Type of this AST node, derived from grammar e.g. `expr`, `factor`, `term`.
+    pub ty: U,
+    /// Level of indentation, useful for pretty printing, formatting, etc.
+    pub indent: u16,
```

Comment:
I agree that `u16` *might* be too small. I think `u32` is quite big enough for this, and it's better to use less memory than more :)